test/test-test.js

```diff
@@ -233,6 +233,69 @@
             }).call({}, "arg1", {}, done);
         },
 
+        "async promise-based test": {
+            requiresSupportFor: {
+                Promise: typeof Promise !== "undefined"
+            },
+
+            "restores the sandbox when the promise is fulfilled": function (done) {
+                var method = function () {};
+                var object = { method: method };
+                var spy = sinon.spy();
+                sinon.test(function (callback) {
+                    var stub;
+                    var sandbox = this;
+                    return Promise.resolve()
+                        .then(function () {
```

Comment:
If you stub `sandbox.restore`, wouldn't that be the assertion you want? You're confirming the side effects of restoring now instead of the actual functionality you want to confirm.